rdar://problem/26398265: Sema: Don't warn on casts that change an ObjC generic's parameters.

jckarter opened the pull request:
Some ObjC generic APIs rely on covariance, and until we devise a systematic solution for SR-1522, you need to force-cast to get the types to line up in cases like this:

func foo(x: GenericClass<NSMutableString>) {
  let x2 = x as! GenericClass<NSString>
  takeGenericClassOfNSString(x2)
}

We currently raise a "cast will never succeed" warning, which is wrong because the dynamic cast can't even check the generic parameters. Suppress this warning. Fixes rdar://problem/26398265.

jckarter commented:
@swift-ci Please test

jckarter commented:
@DougGregor How does this look for Swift 3?

DougGregor commented on lib/Sema/TypeCheckConstraints.cpp:
```diff
@@ -2804,14 +2804,25 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
+  // classes. This may be necessary to force-fit ObjC APIs that depend on
+  // covariance, or for APIs where the generic parameter annotations in the
+  // ObjC headers are inaccurate.
+  if (clas && clas->isObjC() && clas->isGenericContext()) {
```
I think you want to also check clas->hasClangNode(), because you're asking whether it's using the ObjC lightweight generics model.

jckarter replied:
Good point, thanks!

jckarter commented:
Linux failure looks unrelated, but let me address Doug's review comment before merging.

tkremenek commented:
@swift-ci test linux

jckarter commented:
@DougGregor I addressed your review. As you theorized, we did in fact get the dynamic cast behavior wrong in the optimizer, which I also fixed. Would you be willing to review this new set of patches?

jckarter commented:
@swift-ci Please test

tkremenek commented:
@swift-ci test

tkremenek commented:
@swift-ci test

DougGregor commented on lib/SIL/DynamicCasts.cpp:
```diff
@@ -367,6 +367,37 @@ swift::classifyDynamicCast(Module *M,
+            auto sourceSuper = sourceClass;
+            while ((sourceSuper = sourceSuper->getSuperclassDecl())) {
+              if (sourceSuper == targetClass)
+                return DynamicCastFeasibility::WillSucceed;
```
It sure is a pity to open-code this same loop twice in a row...

jckarter replied:
Still passes the "1, 2, or N" test…

DougGregor commented:
Aside from one bit of duplicated code, LGTM!

jckarter commented:
Looks like I raced with some `#include` churn in the last CI run. Let me update and try again.

jckarter commented:
@swift-ci Please test

DougGregor commented:
De-dupe looks good to me, thanks!
